### PR TITLE
Add FRC Driver Station connection support.

### PR DESCRIPTION
**Pull request by PeterJohnson**
The 2017 FRC Driver Station supports getting the robot IP via a TCP
connection that returns JSON.  Use this to support overriding the
server IP address used for client connections.

Default to using this approach for client connections in the NetworkTable
interfaces.

Also add support for setting the server address without stopping and
restarting the client.

### CODE REVIEW

File: include/ntcore_c.h [hunk 1]
```diff
@@ -363,6 +363,11 @@ void NT_StartServer(const char* persist_filename, const char* listen_address,
 void NT_StopServer(void);
 
 /** Starts Client
+ * Starts a client.  Use NT_SetServer to set the server name and port.
+ */
+void NT_StartClientNone(void);
```
Comment (virtuald) [inline include/ntcore_c.h:368]: In pynetworktables I have a test mode that probably does the same thing as this

File: src/networktables/NetworkTable.cpp [hunk 2]
```diff
@@ -44,23 +41,41 @@ void NetworkTable::SetClientMode() { s_client = true; }
...
-#ifdef _MSC_VER
-  sprintf_s(tmp, "roboRIO-%d-FRC.local", team);
-#else
-  using namespace std;
-  snprintf(tmp, 30, "roboRIO-%d-FRC.local", team);
-#endif
-  SetIPAddress(tmp);
+  std::pair<StringRef, unsigned int> servers[3];
+
+  // roboRIO-<team>-FRC.local
```
Comment (virtuald) [inline src/networktables/NetworkTable.cpp]: Is there a preferred ordering or are they done in parallel? If ordered, I'd prefer the IP address to be first.

File: src/networktables/NetworkTable.cpp [hunk 3]
```diff
@@ -20,11 +20,8 @@ unsigned int NetworkTable::s_port = NT_DEFAULT_PORT;
 void NetworkTable::Initialize() {
   if (s_running) Shutdown();
   if (s_client) {
-    std::vector<std::pair<StringRef, unsigned int>> servers;
-    servers.reserve(s_ip_addresses.size());
-    for (const auto& ip_address : s_ip_addresses)
-      servers.emplace_back(std::make_pair(ip_address, s_port));
-    nt::StartClient(servers);
+    nt::StartClient();
+    nt::SetServerFromDS(s_port);
```
Comment (virtuald) [inline src/networktables/NetworkTable.cpp]: Should this be done automatically? It feels like it's something that should be enabled explicitly?

File: src/Dispatcher.cpp [hunk 4]
```diff
@@ -195,6 +199,27 @@ void DispatcherBase::NotifyConnections(
...
+  connectors.push_back(connector);
+  SetConnector(std::move(connectors));
+}
+
+void DispatcherBase::SetConnector(std::vector<Connector>&& connectors) {
+  std::lock_guard<std::mutex> lock(m_user_mutex);
+  m_client_connectors = std::move(connectors);
+}
+
+void DispatcherBase::SetConnectorOverride(Connector connector) {
```
Comment (virtuald) [inline src/Dispatcher.cpp:213]: This connector override thing seems strange to me. Why not just query all of the possible connectors for potential addresses to connect to?

File: src/networktables/NetworkTable.cpp (hunk 2 above)
Comment (PeterJohnson) [reply src/networktables/NetworkTable.cpp]: It's round robin, so all will get tried.

File: src/Dispatcher.cpp (hunk 4 above)
Comment (PeterJohnson) [reply src/Dispatcher.cpp:213]: The idea is for the DS discovered IP to not completely clobber the user-set server list, so when the DS disconnects, we fall back to trying the user list.

File: include/ntcore_c.h (hunk 1 above)
Comment (PeterJohnson) [reply include/ntcore_c.h:368]: This really isn't a test mode.  It's just allowing separation of starting up the client and setting the server addresses.

**Pull request merged**